sqldb/migrations_test.go

```diff
@@ -0,0 +1,170 @@
+package sqldb
+
+import (
+	"context"
+	"testing"
+
+	"github.com/lightningnetwork/lnd/sqldb/sqlc"
+	"github.com/stretchr/testify/require"
+)
+
+// makeMigrationTestDB is a type alias for a function that creates a new test
+// database and returns the base database and a function that executes selected
+// migrations.
+type makeMigrationTestDB = func(*testing.T, uint) (*BaseDB,
+	func(MigrationTarget) error)
+
+// TestMigrations is a meta test runner that runs all migration tests.
+func TestMigrations(t *testing.T) {
+	sqliteTestDB := func(t *testing.T, version uint) (*BaseDB,
+		func(MigrationTarget) error) {
+
+		db := NewTestSqliteDBWithVersion(t, version)
+
+		return db.BaseDB, db.ExecuteMigrations
+	}
+
+	postgresTestDB := func(t *testing.T, version uint) (*BaseDB,
+		func(MigrationTarget) error) {
+
+		pgFixture := NewTestPgFixture(t, DefaultPostgresFixtureLifetime)
+		t.Cleanup(func() {
+			pgFixture.TearDown(t)
+		})
+
+		db := NewTestPostgresDBWithVersion(
+			t, pgFixture, version,
+		)
+
+		return db.BaseDB, db.ExecuteMigrations
+	}
+
+	tests := []struct {
+		name string
+		test func(*testing.T, makeMigrationTestDB)
+	}{
+		{
+			name: "TestInvoiceExpiryMigration",
+			test: testInvoiceExpiryMigration,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+
+		t.Run(test.name+"_SQLite", func(t *testing.T) {
+			test.test(t, sqliteTestDB)
+		})
+
+		t.Run(test.name+"_Postgres", func(t *testing.T) {
+			test.test(t, postgresTestDB)
+		})
+
+	}
+}
+
+// TestInvoiceExpiryMigration tests that the migration from version 3 to 4
+// correctly sets the expiry value of normal invoices to 86400 seconds and
+// 2592000 seconds for AMP invoices.
+func testInvoiceExpiryMigration(t *testing.T, makeDB makeMigrationTestDB) {
+	t.Parallel()
+	ctxb := context.Background()
+
+	// Create a new database that already has the first version of the
+	// native invoice schema.
+	db, migrate := makeDB(t, 3)
+
+	// Add a few invoices. For simplicity we reuse the payment hash as the
+	// payment address and payment request hash instead of setting them to
+	// NULL (to not run into uniqueness constraints). Note that SQLC
+	// currently doesn't support nullable blob fields porperly. A workaround
+	// is in progress: https://github.com/sqlc-dev/sqlc/issues/3149
+
+	// Add an invoice where is_amp will be set to NULL.
+	hash1 := []byte{1, 2, 3}
+	_, err := db.InsertInvoice(ctxb, sqlc.InsertInvoiceParams{
+		Hash:               hash1,
+		PaymentAddr:        hash1,
+		PaymentRequestHash: hash1,
+		Expiry:             -123,
```

Comment:
does this actually set it to NULL? surely it just sees this as `IsAmp: false`?

Reply:
Great catch, fixed!